File: backend/app/llm_adapter.py
import os
import json
import uuid
from typing import List

PROVIDER = os.environ.get("LLM_PROVIDER", "mock")


class MockAdapter:
    def extract_tasks(self, transcript_text: str) -> List[dict]:
        # Very simple mock: split transcript into lines and produce tasks
        lines = [l.strip() for l in transcript_text.splitlines() if l.strip()]
        tasks = []
        # create up to 5 tasks
        for i, line in enumerate(lines[:5]):
            tid = f"t{i+1}"
            tasks.append(
                {
                    "id": tid,
                    "description": line[:200],
                    "priority": "medium" if i % 2 == 0 else "low",
                    "dependencies": [f"t{i}"] if i > 0 else [],
                    "status": "ready",
                }
            )
        # Intentionally include a hallucinated dependency to test sanitization
        if tasks:
            tasks[0]["dependencies"].append("unknown_task_id")
        return tasks


class OpenAIAdapter:
    def __init__(self, api_key: str, model: str = None):
        try:
            import openai
        except Exception as e:
            raise RuntimeError("openai package is not installed. Install it to use OpenAIAdapter") from e
        self.openai = openai
        self.api_key = api_key
        self.model = model or os.environ.get("OPENAI_MODEL", "gpt-4o-mini")
        self.openai.api_key = api_key

    def _build_prompt(self, transcript: str) -> str:
        return (
            "You are an assistant that extracts actionable tasks from meeting transcripts. "
            "Return ONLY a JSON array of task objects. No markdown, no commentary.\n\n"
            "Task schema for each item:\n"
            '{ "id": "string", "description": "string", "priority": "low|medium|high", '
            '"dependencies": ["task-id-1", ...], "status": "ready|blocked" }\n\n'
            "Transcript:\n\n"
            f"{transcript}\n\n"
            "Respond with a JSON array only."
        )

    def extract_tasks(self, transcript_text: str) -> List[dict]:
        prompt = self._build_prompt(transcript_text)
        resp = self.openai.ChatCompletion.create(
            model=self.model,
            messages=[{"role": "user", "content": prompt}],
            temperature=0.0,
        )
        content = ""
        try:
            # compatible with various OpenAI response shapes
            content = resp["choices"][0]["message"]["content"]
        except Exception:
            content = resp.get("choices", [{}])[0].get("text", "")
        return _parse_json_from_string(content)
def _parse_json_from_string(s: str) -> List[dict]:
    s = s.strip()
    # Try direct parse
    try:
        parsed = json.loads(s)
        if isinstance(parsed, list):
            return parsed
        # if the model returned an object with a top-level 'tasks' key
        if isinstance(parsed, dict) and "tasks" in parsed and isinstance(parsed["tasks"], list):
            return parsed["tasks"]
    except Exception:
        pass
    # Fallback: extract first JSON array substring
    import re

    m = re.search(r"(\[\\s*\\{[\\s\\S]*?\\}\\s*\\])", s)
    if m:
        try:
            return json.loads(m.group(1))
        except Exception:
            pass
    # try find first '[' and last ']' and parse between
    try:
        start = s.index("[")
        end = s.rindex("]") + 1
        return json.loads(s[start:end])
    except Exception:
        pass
    # give up defensively
    return []


def get_adapter():
    prov = os.environ.get("LLM_PROVIDER", PROVIDER).lower()
    if prov == "openai":
        key = os.environ.get("OPENAI_API_KEY") or os.environ.get("LLM_API_KEY")
        if not key:
            raise RuntimeError("OPENAI_API_KEY or LLM_API_KEY required for OpenAI provider")
        return OpenAIAdapter(api_key=key)
    # default to mock
    return MockAdapter()

File: backend-ts/src/llm_adapter.ts
import { v4 as uuidv4 } from "uuid";
import { Configuration, OpenAIApi } from "openai";

export type TaskItem = {
  id: string;
  description: string;
  priority: "low" | "medium" | "high";
  dependencies: string[];
  status: "ready" | "blocked";
};

export class MockAdapter {
  async extractTasks(transcript: string): Promise<TaskItem[]> {
    const lines = transcript.split("\n").map((l) => l.trim()).filter(Boolean);
    const tasks: TaskItem[] = [];
    for (let i = 0; i < Math.min(lines.length, 6); i++) {
      const tid = `t${i + 1}`;
      tasks.push({
        id: tid,
        description: lines[i].slice(0, 200),
        priority: i % 2 === 0 ? "medium" : "low",
        dependencies: i > 0 ? [`t${i}`] : [],
        status: "ready"
      });
    }
    if (tasks.length) {
      // include a hallucinated dependency to test sanitization
      tasks[0].dependencies.push("unknown_task_id");
    }
    return tasks;
  }
}

export class OpenAIAdapter {
  client: OpenAIApi;
  model: string;
  constructor(apiKey: string) {
    const conf = new Configuration({ apiKey });
    this.client = new OpenAIApi(conf);
    this.model = process.env.OPENAI_MODEL || "gpt-4o-mini";
  }

  async extractTasks(transcript: string): Promise<TaskItem[]> {
    const prompt = `You are an assistant that extracts actionable tasks from meeting transcripts. Return ONLY a JSON array of task objects following this schema: { "id": "string", "description": "string", "priority": "low|medium|high", "dependencies": ["task-id-1"], "status": "ready|blocked" }. Transcript:\n\n${transcript}\n\nReturn JSON array only.`;
    const resp = await this.client.createChatCompletion({
      model: this.model,
      messages: [{ role: "user", content: prompt }],
      temperature: 0.0,
      max_tokens: 1500,
    } as any);
    let content = "";
    try {
      content = (resp.data.choices?.[0] as any)?.message?.content ?? (resp.data.choices?.[0] as any)?.text ?? "";
    } catch (e) {
      content = "";
    }
    // try parse
    try {
      const parsed = JSON.parse(content);
      if (Array.isArray(parsed)) return parsed;
      if (parsed?.tasks && Array.isArray(parsed.tasks)) return parsed.tasks;
    } catch {
      // fallback find first array
      const m = content.match(/\[[\s\S]*\]/);
      if (m) {
        try {
          return JSON.parse(m[0]);
        } catch {}
      }
    }
    return [];
  }
}

export function getAdapter() {
  const provider = (process.env.LLM_PROVIDER || "mock").toLowerCase();
  if (provider === "openai") {
    const key = process.env.OPENAI_API_KEY || process.env.LLM_API_KEY;
    if (!key) throw new Error("OPENAI_API_KEY required for openai provider");
    return new OpenAIAdapter(key);
  }
  return new MockAdapter();
}

